File: packages/hmssdk_flutter/android/src/main/kotlin/live/hms/hmssdk_flutter/methods/HMSNoiseCancellationControllerAction.kt
package live.hms.hmssdk_flutter.methods

import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import live.hms.hmssdk_flutter.HMSCommonAction
import live.hms.hmssdk_flutter.HMSErrorLogger
import live.hms.hmssdk_flutter.HMSResultExtension
import live.hms.video.factories.noisecancellation.AvailabilityStatus
import live.hms.video.sdk.HMSSDK

class HMSNoiseCancellationControllerAction {
    companion object {
        fun noiseCancellationActions(
            call: MethodCall,
            result: MethodChannel.Result,
            hmssdk: HMSSDK,
        ) {
            when (call.method) {
                "enable_noise_cancellation" -> {
                    enable(result, hmssdk)
                }
                "disable_noise_cancellation" -> {
                    disable(result, hmssdk)
                }
                "is_noise_cancellation_enabled" -> {
                    isEnabled(result, hmssdk)
                }
                "is_noise_cancellation_available" -> {
                    isAvailable(result, hmssdk)
                }
                else -> {
                    result.notImplemented()
                }
            }
        }

        /**
         * [enable] method enables noise cancellation for the user
         */
        private fun enable(
            result: MethodChannel.Result,
            hmssdk: HMSSDK,
        ) {
            hmssdk.enableNoiseCancellation(true,HMSCommonAction.getActionListener(result))
        }

        /**
         * [disable] method disables noise cancellation for the user
         */
        private fun disable(
            result: MethodChannel.Result,
            hmssdk: HMSSDK,
        ) {
            hmssdk.enableNoiseCancellation(false, HMSCommonAction.getActionListener(result))
        }

        /**
         * [isEnabled] method returns whether noise cancellation is enabled or not
         */
        private fun isEnabled(
            result: MethodChannel.Result,
            hmssdk: HMSSDK,
        ) {
            val isEnabled = hmssdk.isNoiseCancellationEnabled()
            result.success(HMSResultExtension.toDictionary(true, isEnabled))
        }

        /**
         * [isAvailable] method returns whether noise cancellation is available in the room
         */
        private fun isAvailable(
            result: MethodChannel.Result,
            hmssdk: HMSSDK,
        ) {
            val availabilityStatus = hmssdk.isNoiseCancellationSupported()
            if (availabilityStatus == AvailabilityStatus.Available) {
                result.success(HMSResultExtension.toDictionary(true, data = true))
            } else {
                val reason = (availabilityStatus as AvailabilityStatus.NotAvailable).reason
                HMSErrorLogger.logError("isAvailable", reason, "NoiseCancellation Error")
                result.success(HMSResultExtension.toDictionary(true, data = false))
            }
        }
    }
}

File: packages/hmssdk_flutter/ios/Classes/Actions/HMSNoiseCancellationControllerAction.swift
//
//  HMSNoiseCancellationController.swift
//  hmssdk_flutter
//
//  Created by Pushpam on 22/03/24.
//

import Foundation
import HMSSDK
import HMSNoiseCancellationModels

class HMSNoiseCancellationController {
    static var noiseCancellationController: HMSNoiseCancellationPlugin?

    static func noiseCancellationActions(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        switch call.method {
        case "enable_noise_cancellation":
            enable(result)
        case "disable_noise_cancellation":
            disable(result)
        case "is_noise_cancellation_enabled":
            isEnabled(result)
        case "is_noise_cancellation_available":
            isAvailable(result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    /**
     * [createPlugin] creates the noiseCancellationPlugin based on the parameter in audio track settings
     */
    static func createPlugin() {
        noiseCancellationController = {
            if let pathForNCModel = HMSNoiseCancellationModels.path(for: .smallFullBand) {
                 return HMSNoiseCancellationPlugin(modelPath: pathForNCModel, initialState: .enabled)
            } else {
                 assertionFailure("Noise cancellation model was not found")
            }
            return nil
       }()
    }

    /**
     * [enable] method enables noise cancellation for the user
     */
    private static func enable(_ result: @escaping FlutterResult) {
        if let controller = noiseCancellationController {
            do {
                try controller.enable()
                result(nil)
            } catch {
                print("An error occurred: \(error)")
            }
        } else {
            HMSErrorLogger.logError(#function, "Noise cancellation controller is not initialised. Please enable noise cancellation in audio track settings while initialising HMSSDK", "NULL_ERROR")
        }
    }

    /**
     * [disable] method disables noise cancellation for the user
     */
    private static func disable(_ result: @escaping FlutterResult) {
        if let controller = noiseCancellationController {
            do {
                try controller.disable()
                result(nil)
            } catch {
                print("An error occurred: \(error)")
            }
        } else {
            HMSErrorLogger.logError(#function, "Noise cancellation controller is not initialised. Please enable noise cancellation in audio track settings while initialising HMSSDK", "NULL_ERROR")
        }
    }

    /**
     * [isEnabled] method returns whether noise cancellation is enabled or not
     */
    private static func isEnabled(_ result: @escaping FlutterResult) {
        if let controller = noiseCancellationController {
                let isNoiseCancellationEnabled = controller.isEnabled()
            result(HMSResultExtension.toDictionary(true, isNoiseCancellationEnabled))
        } else {
            HMSErrorLogger.logError(#function, "Noise cancellation controller is not initialised. Please enable noise cancellation in audio track settings while initialising HMSSDK", "NULL_ERROR")
        }
    }

    /**
     * [isAvailable] method returns whether noise cancellation is available in the room
     */
    private static func isAvailable(_ result: @escaping FlutterResult) {
        if let controller = noiseCancellationController {
            let isNoiseCancellationAvailable = controller.isNoiseCancellationAvailable
            result(HMSResultExtension.toDictionary(true, isNoiseCancellationAvailable))
        } else {
            HMSErrorLogger.logError(#function, "Noise cancellation controller is not initialised. Please enable noise cancellation in audio track settings while initialising HMSSDK", "NULL_ERROR")
        }
    }

}
